### PR TITLE
fix #1970 - update MultiPartFormDataContent to allow contentType override using optional builder

### LINKED ISSUE #1970: Objections to changing boundary to internal on MultiPartFormDataContent?
https://github.com/ktorio/ktor/blob/3ce3906c2bfa3c86238d40e9b67006b2b020bdaf/ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt#L43

I'd like to change this to internal so I can consume the boundary like so:

```kotlin
class MultiPartMixedDataContent(
        parts: List<PartData>
) : MultiPartFormDataContent {
    override val contentType: ContentType = ContentType.MultiPart.Mixed.withParameter("boundary", boundary)
}
```

### PR DESCRIPTION
**Pull request by jschneidereit**
Fixes #1970 

**Subsystem**
Client

**Motivation**
Before this change `MultiPartFormDataContent` was not compatible with servers that accepted `multipart/mixed` content. 

**Solution**
Now a user can pass their own lambda for building content type like so:

```kotlin
val mixed = MultiPartFormDataContent(
    parts = listOf(),
    buildContentType = { b -> ContentType.MultiPart.Mixed.withParameter("boundary", b) }
)
```

### CODE REVIEW

File: ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt [hunk 1]
```diff
@@ -31,14 +31,18 @@ class FormDataContent(
...
 
 /**
  * [OutgoingContent] for multipart/form-data formatted request.
  *
  * @param parts: form part data
  */
 class MultiPartFormDataContent(
-    parts: List<PartData>
+    parts: List<PartData>,
+    buildContentType: (boundary: String) -> ContentType = buildFormDataContentType
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt:45]: Could we move `boundary` and `contentType` fields in the constructor?

**Comment by e5l** [comment]
Hey @jschneidereit, sorry for the delay. Could you check one thing before the merge?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt (hunk 1 above)
Comment (jschneidereit) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt]: @e5l sure! pushed the fix!

**Comment by jschneidereit** [comment]
> Hey @jschneidereit, sorry for the delay. Could you check one thing before the merge?

@e5l did I miss something?

File: ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt [hunk 2]
```diff
@@ -31,16 +31,16 @@ public class FormDataContent(
...
-
 /**
  * [OutgoingContent] for multipart/form-data formatted request.
  *
  * @param parts: form part data
  */
 public class MultiPartFormDataContent(
-    parts: List<PartData>
+    parts: List<PartData>,
+    val boundary: String = generateBoundary(),
```
Comment (e5l) [inline ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt:41]: Could you tell me why do we need `boundary` in the constructor?

**Review by e5l — CHANGES_REQUESTED** [review]

File: ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt (hunk 2 above)
Comment (jschneidereit) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt:41]: Hi @e5l I had originally made it override-able but @cy6erGn0m requested an optional constructor argument in the discussion in #1970

Comment (jschneidereit) [reply ktor-client/ktor-client-core/common/src/io/ktor/client/request/forms/FormDataContent.kt:41]: The other option to only having `contentType` in the constructor and not `boundary` would be a lambda that accepts a string like `contentTypeFn: (String) -> ContentType` but I think that's a little overkill.

**Review by e5l — APPROVED** [review]

**Pull request merged**